components/net/fetch/methods.rs

```diff
@@ -866,6 +869,405 @@ async fn scheme_fetch(
     }
 }
 
+fn consider_local_directory_listing(
+    request: &mut Request,
+    url: ServoUrl,
+    file_path: std::path::PathBuf,
+) -> Response {
+    /*
+    If you want to be able to browse local directories, configure
+    Servo prefs so that "browser.local_directory_listing.enabled"
+    is set to true.
+    */
+    if pref!(browser.local_directory_listing.enabled) {
+        display_local_directory_listing(request, url, file_path)
+    } else {
+        Response::network_error(NetworkError::Internal(
+            "Local directory listing feature has not been enabled in preferences".into(),
+        ))
+    }
+}
+
+fn display_local_directory_listing(
```

Comment:
I don't suppose it's possible to put the unit tests directly into the source code file, is it?

At the moment, the only way I've been able to make the methods visible to the unit tests is to mark them with `pub` because `pub(crate)` does not work (even though they're both under the same crate, to my understanding of the term "crate").

If having unit tests inside the source file is not acceptable in the Servo project, do you know how I can limit visibility of the local directory listing methods so that they're visible to the unit tests without being fully public?

Reply:
As mentioned in the Servo Zulip forum ("Directory listing feature" thread), I don't believe it's possible to unit test methods without making them fully public, because the unit tests appear under a directory with a `main.rs` file, which marks it out as a binary crate. While the main code lives under a directory with a `lib.rs` file, marking it out as a library crate. So the `pub(crate)` modifier will not make the library crate code visible to the unit tests which are in a separate (binary) crate.

For now I'm just copying the pattern used elsewhere, and I've made the methods and structures `pub` so that they can be unit tested.